README.md
# chartjs-ror

Simplifies using [Chart.js][] in Rails views.


## Current Chart.js version

This gem includes [Chart.js v2.7.2](https://github.com/chartjs/Chart.js/tree/v2.7.2).


## Installation

Add this line to your application's Gemfile:

    gem 'chartjs-ror'

And then execute:

    $ bundle

There are two builds of Chart.js: with or without Moment.js.

Choose the appropriate one in your javascript manifest:

- Without Moment.js:
  ```javascript
  //= require Chart.min
  ```

- With Moment.js:
  ```javascript
  //= require Chart.bundle.min
  ```

Please note Chart.js [no longer supports IE8 and below](https://github.com/chartjs/Chart.js/issues/2396#issuecomment-215233106).


## Usage

Each chart type has a corresponding helper for your views.  The helper methods take the same arguments as their Javascript counterparts.  The `options` are optional.


```erb
<%= line_chart           data, options %>
<%= bar_chart            data, options %>
<%= horizontal_bar_chart data, options %>
<%= radar_chart          data, options %>
<%= polar_area_chart     data, options %>
<%= pie_chart            data, options %>
<%= doughnut_chart       data, options %>
<%= bubble_chart         data, options %>
<%= scatter_chart        data, options %>
```

If you don't want these helpers – perhaps they clash with other methods in your views – add this initializer to your app:

```ruby
# config/initializers/chartjs.rb
Chartjs.no_conflict!
```

Then you can use these helpers instead:

```erb
<%= chartjs_line_chart           data, options %>
<%= chartjs_bar_chart            data, options %>
<%= chartjs_horizontal_bar_chart data, options %>
<%= chartjs_radar_chart          data, options %>
<%= chartjs_polar_area_chart     data, options %>
<%= chartjs_pie_chart            data, options %>
<%= chartjs_doughnut_chart       data, options %>
<%= chartjs_bubble_chart         data, options %>
<%= chartjs_scatter_chart        data, options %>
```

For example, to render a [line chart][linechart] in Javascript:

```javascript
var data = {
    labels: ["January", "February", "March", "April", "May", "June", "July"],
    datasets: [
        {
            label: "My First dataset",
            backgroundColor: "rgba(220,220,220,0.2)",
            borderColor: "rgba(220,220,220,1)",
            data: [65, 59, 80, 81, 56, 55, 40]
        },
        {
            label: "My Second dataset",
            backgroundColor: "rgba(151,187,205,0.2)",
            borderColor: "rgba(151,187,205,1)",
            data: [28, 48, 40, 19, 86, 27, 90]
        }
    ]
};
var options = { ... };
new Chart(ctx, {
    type: 'Line',
    data: data,
    options: options
});
```

The Ruby equivalent is:

```ruby
data = {
  labels: ["January", "February", "March", "April", "May", "June", "July"],
  datasets: [
    {
        label: "My First dataset",
        backgroundColor: "rgba(220,220,220,0.2)",
        borderColor: "rgba(220,220,220,1)",
        data: [65, 59, 80, 81, 56, 55, 40]
    },
    {
        label: "My Second dataset",
        backgroundColor: "rgba(151,187,205,0.2)",
        borderColor: "rgba(151,187,205,1)",
        data: [28, 48, 40, 19, 86, 27, 90]
    }
  ]
}
options = { ... }
<%= line_chart data, options %>
```

You can also use underscored symbols for keys, instead of the camelcase versions.
They will be converted to their lower camelcase counterparts on output.

```ruby
data = {
  labels: ["January", "February", "March", "April", "May", "June", "July"],
  datasets: [
    {
        label: "My First dataset",
        background_color: "rgba(220,220,220,0.2)",
        border_color: "rgba(220,220,220,1)",
        data: [65, 59, 80, 81, 56, 55, 40]
    },
    {
        label: "My Second dataset",
        background_color: "rgba(151,187,205,0.2)",
        border_color: "rgba(151,187,205,1)",
        data: [28, 48, 40, 19, 86, 27, 90]
    }
  ]
}
options = { ... }
<%= line_chart data, options %>
```

### Options

You can put anything in the `options` hash that Chart.js recognises.  To pass a JavaScript function as an option value, wrap it in quotation marks to make it a string.

You can also use these non-Chart.js settings:

* `:class`          - class of the DOM canvas element - default is `chart`.
* `:id`             - id of the DOM canvas element - default is `chart-n` where `n` is the 0-based index of the chart on the page.
* `:width`          - width of the canvas in px - default is `400`.
* `:height`         - height of the canvas in px - default is `400`.


## Sample output

```html
<canvas id="chart-0" class="chart" width=400 height=400></canvas>
<script type="text/javascript">
  //<![CDATA[
  (function() {
    var initChart = function() {
      var ctx = document.getElementById("chart-0");
      var chart = new Chart(ctx, {
        type: "Line",
        data = { ... };
        options = { ... };
      });
    };

    if (typeof Chart !== "undefined" && Chart !== null) {
      initChart();
    }
    else {
      if (window.addEventListener) {
        window.addEventListener("load", initChart, false);
      }
      else if (window.attachEvent) {
        window.attachEvent("onload", initChart);
      }
    }
  })();
  //]]>
</script>
```

The Javascript is actually written out on a single line but you get the idea.


## Intellectual Property

Copyright Andrew Stewart, AirBlade Software.  Released under the MIT licence.


  [Chart.js]: http://www.chartjs.org/
  [linechart]: http://www.chartjs.org/docs/#line-chart



CHANGELOG
# CHANGELOG


## 3.6.2 - 2018-03-14

- Use Chart.js v2.7.2.


## 3.6.1 - 2018-03-14

- Use Chart.js v2.7.1.


## 3.6.0 - 2018-03-14

- Use Chart.js v2.7.0.


## 3.5.0 - 2018-03-14

- Use Chart.js v2.6.0.


## 3.4.0 - 2018-03-14

- Use Chart.js v2.5.0.


## 3.3.1 - 2017-01-27

- Improve conversion of data and options hashes to JavaScript.


## 3.3.0 - 2017-01-26

- Use Chart.js v2.4.0.


## 3.2.0 - 2017-01-25

- Use Chart.js v2.3.0.


## 3.1.0 - 2017-01-25

- Use Chart.js v2.2.0.
- Add CHANGELOG.


## 3.0.2 - 2016-06-13

- Add pie chart helper.


## 3.0.1 - 2016-06-10

- Add horizontal bar chart helper.


## 3.0.0 - 2016-06-08

- Use Chart.js v2.1.4 (backwards incompatible).


## 2.2.0 - 2015-08-26

- Use Chart.js v1.0.2.


## 2.1.3 - 2015-08-26

- AJAX and Turbolinks compatibility.


## 2.1.2 - 2015-08-18

- Pass `onAnimationComplete` and `onAnimationProgress` functions correctly.


## 2.1.1 - 2015-03-13

- Provide original helpers with easy way to opt out.
- Namespace the view helpers.
- Fix name of polar_area_chart in README.
- Ensure any old charts are cleaned up.
- Clarify optional features.
- State Chart.js version in README.


## 2.1.0 - 2015-01-14

- Use Chart.js v1.0.1.
- Support Turbolinks.
- Do not generate a legend by default.


## 2.0.0 - 2014-11-07

- Use Chart.js 1cfa42bc (a little after v1.0.1-beta.4).
- Wrap generated JavaScript snippet in a closure.
- Use Chart.js' legends.
- Update README.
- Add link to scale calculation issue.


## 1.1.1 - 2014-03-20

- Rails 4 compatibility.


## 1.1.0 - 2014-03-14

- Use Rails' asset pipeline.


## 1.0.1 - 2013-06-21

- Opt-in alternative scale calculations.
- Code for filling gaps in series.


## 1.0.0 - 2013-06-16

- First release.



test/test_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)

require 'minitest/autorun'


test/axis_helpers_test.rb
require 'minitest/autorun'
require 'chartjs/axis_helpers'

class AxisHelpersTest < MiniTest::Unit::TestCase

  def setup
    @subject = Object.new
    @subject.extend Chartjs::AxisHelpers
  end

  def test_series_empty_data
    data = {}
    result = @subject.series data
    assert_equal 0, result.length
  end

  def test_series
    data = { 2 => 42 }
    result = @subject.series data
    assert_equal 3, result.length
    assert_equal({0 => 0, 1 => 0, 2 => 42}, result)
  end

  def test_calculate_scale_max
    {
         0 =>    0,
         3 =>    5,
         5 =>    5,
         6 =>   10,
         9 =>   10,
        10 =>   10,
        12 =>   15,
        16 =>   20,
        60 =>   60,
        65 =>   65,
       601 =>  650,
      1430 => 1500,
      7654 => 8000,
    }.each do |input, expected_output|
      assert_equal expected_output, @subject.calculate_scale_max(input)
    end
  end

  def test_ordinate_scale
    data = []
    assert_equal({scaleSteps: 0, scaleStepWidth: 0, scaleStartValue: 0}, @subject.ordinate_scale(data))

    data = [ 12 ]
    assert_equal({scaleSteps: 5, scaleStepWidth: 3, scaleStartValue: 0}, @subject.ordinate_scale(data))

    data = [ 33 ]
    assert_equal({scaleSteps: 7, scaleStepWidth: 5, scaleStartValue: 0}, @subject.ordinate_scale(data))

    data = [ 601 ]
    assert_equal({scaleSteps: 7, scaleStepWidth: 100, scaleStartValue: 0}, @subject.ordinate_scale(data))

    data = [ 999 ]
    assert_equal({scaleSteps: 5, scaleStepWidth: 200, scaleStartValue: 0}, @subject.ordinate_scale(data))

    data = [ 13001 ]
    assert_equal({scaleSteps: 5, scaleStepWidth: 3000, scaleStartValue: 0}, @subject.ordinate_scale(data))
  end

end


test/chart_helpers_test.rb
require 'test_helper'

require 'chartjs/chart_helpers'

class ChartHelpersTest < Minitest::Test

  class Foo
    include Chartjs::ChartHelpers

    def to_js(element)
      to_javascript_string element
    end
  end

  def setup
    @subject = Foo.new
  end

  def test_to_javascript_string
    assert_equal '"foo"', @subject.to_js('foo')

    assert_equal '["a",42]', @subject.to_js(['a', 42])

    assert_equal '{"a":42}', @subject.to_js({'a' => 42})

    assert_equal '{"a":function(x) { blah; }}', @subject.to_js({'a' => 'function(x) { blah; }'})

    expected = \
'{"a":function(x) {
  if (x) {
    foo();
  }
  else {
    bar();
  }
}
}'

    assert_equal expected, @subject.to_js({'a' => <<END})
function(x) {
  if (x) {
    foo();
  }
  else {
    bar();
  }
}
END
  end

end


lib/chartjs/chart_helpers.rb
require 'json'
require 'bigdecimal'

module Chartjs
  module ChartHelpers

    CHART_TYPES = %w[ bar bubble doughnut horizontal_bar line pie polar_area radar scatter ]

    module Explicit
      CHART_TYPES.each do |type|
        define_method "chartjs_#{type}_chart" do |data, options = {}|    # def chartjs_polar_area_chart(data, options = {})
          chart type, data, options                                      #   chart 'polar_area', data, options
        end                                                              # end
      end
      include Chartjs::ChartHelpers
    end

    module Implicit
      CHART_TYPES.each do |type|
        define_method "#{type}_chart" do |data, options = {}|            # def polar_area_chart(data, options = {})
          chart type, data, options                                      #   chart 'polar_area', data, options
        end                                                              # end
      end
      include Chartjs::ChartHelpers
    end

    private

    def chart(type, data, options)
      opts = options.dup

      @chart_id ||= -1
      element_id = opts.delete(:id)     || "chart-#{@chart_id += 1}"
      css_class  = opts.delete(:class)  || 'chart'
      width      = opts.delete(:width)  || '400'
      height     = opts.delete(:height) || '400'

      canvas = content_tag :canvas, '', id: element_id, class: css_class, width: width, height: height

      script = javascript_tag do
        <<-END.squish.html_safe
        (function() {
          var initChart = function() {
            var ctx = document.getElementById(#{element_id.to_json});
            var chart = new Chart(ctx, {
              type: "#{camel_case type}",
              data: #{to_javascript_string data},
              opts: #{to_javascript_string opts}
            });
          };

          if (typeof Chart !== "undefined" && Chart !== null) {
            initChart();
          }
          else {
            /* W3C standard */
            if (window.addEventListener) {
              window.addEventListener("load", initChart, false);
            }
            /* IE */
            else if (window.attachEvent) {
              window.attachEvent("onload", initChart);
            }
          }
        })();
        END
      end

      canvas + script
    end

    # polar_area -> polarArea
    def camel_case(string)
      string.gsub(/_([a-z])/) { $1.upcase }
    end

    def to_javascript_string(element)
      case element
      when Hash
        hash_elements = []
        element.each do |key, value|
          hash_elements << camel_case(key.to_s).to_json + ':' + to_javascript_string(value)
        end
        '{' + hash_elements.join(',') + '}'
      when Array
        array_elements = []
        element.each do |value|
          array_elements << to_javascript_string(value)
        end
        '[' + array_elements.join(',') + ']'
      when String
        if element.match(/^\s*function.*}\s*$/m)
          # Raw-copy function definitions to the output without surrounding quotes.
          element
        else
          element.to_json
        end
      when BigDecimal
        element.to_s
      else
        element.to_json
      end
    end

  end
end


lib/chartjs/axis_helpers.rb
require 'active_support'

module Chartjs
  module AxisHelpers

    # Returns an ordered copy of data with "missing" keys mapped to 0 values.
    #
    # data - a Hash of Integer keys mapped to to Integer values.
    #
    # Returns an OrderedHash whose Integer keys run from 0 to data's maximum key
    # and whose values are the corresponding values in data or 0.
    def series(data)
      return {} if data.nil? || data.empty?
      (0..data.keys.max).reduce(ActiveSupport::OrderedHash.new) { |hash,k| hash[k] = data[k] || 0; hash }
    end

    # Calculates the number of steps and the step width for ordinate axis labels.
    # The ordinate axis is assumed to start at 0.
    #
    # data - an Array of Numerics.
    #
    # Returns a Hash with:
    #   :scaleSteps      - the number of steps (always <= 10)
    #   :scaleStepWidth  - the value jump between steps
    #   :scaleStartValue - 0
    def ordinate_scale(data)
      scale_max = calculate_scale_max data.max
      normalised_scale_max = normalise scale_max

      # normalised_scale_max  number_steps  step_width
      #          0                 0           0
      #          0.5               5           1
      #          1                 5           2
      #          1.5               5           3
      #          2                 4           5
      #          2.5               5           5
      #          3                 6           5
      #          3.5               7           5
      #          4                 8           5
      #          4.5               9           5
      #          5                10           5
      #          5.5               6          10
      #          6                 6          10
      #          6.5               7          10
      #          7                 7          10
      #          7.5               8          10
      #          8                 8          10
      #          8.5               9          10
      #          9                 9          10
      #          9.5              10          10
      if normalised_scale_max.zero?
        number_steps = 0
        step_width = 0
      elsif normalised_scale_max < 2
        number_steps = 5
        step_width = scale_max / number_steps
      elsif normalised_scale_max < 5.5
        step_width = 5 * 10**(order_of_magnitude(scale_max) - 1)
        number_steps = scale_max / step_width.to_f
      else
        step_width = 10**(order_of_magnitude(scale_max))
        number_steps = scale_max / step_width.to_f
      end

      {scaleSteps: number_steps.ceil, scaleStepWidth: step_width, scaleStartValue: 0}
    end

    # Rounds up value to the next "increment", where an increment is half
    # the order of magnitude value.
    #
    # value - a Numeric.
    #
    # Examples
    #
    #   input  output
    #     12     15
    #     15     15
    #     16     20
    #   1432   1500
    #   7654   8000
    #
    # Returns an Integer.
    def calculate_scale_max(value)
      normalised = normalise value
      quotient, modulus = normalised.divmod 1
      next_increment = if modulus.zero?
                         quotient
                       elsif modulus <= 0.5
                         quotient + 0.5
                       else
                         quotient + 1
                       end
      (next_increment * 10**(order_of_magnitude(value))).to_i
    end

    # Normalises the value to between 0 inclusive and 10.
    #
    # value - a Numeric.
    #
    # Returns a Float or 0 if value is nil or zero..
    def normalise(value)
      return 0 if value.nil? || value.zero?
      value.to_f / 10**(order_of_magnitude(value))
    end

    # Calculates the order of magnitude of value.
    #
    # value - a Numeric.
    #
    # Examples
    #
    #   order_of_magnitude(3)
    #   # => 1
    #
    #   order_of_magnitude(13)
    #   # => 1
    #
    #   order_of_magnitude(153)
    #   # => 2
    #
    # Returns a Integer or 0 if value is nil or zero.
    def order_of_magnitude(value)
      return 0 if value.nil? || value.zero?
      magnitude = Math.log10 value
      order_of_magnitude = magnitude.floor
      order_of_magnitude.zero? ? 1 : order_of_magnitude
    end

  end
end
